cloud.ganttproject.colloboque/src/main/kotlin/cloud/ganttproject/colloboque/PostgreStorageApi.kt

```diff
@@ -112,6 +117,60 @@ class PostgreStorageApi(private val factory: ConnectionFactory, private val supe
     }
   }
 
+  fun parallelTransactions(
+    projectXml: String,
+    baseTxnId: BaseTxnId,
+    serverTransaction: List<XlogRecord>,
+    clientTransaction: List<XlogRecord>
+  ): Boolean {
+    val settings = Settings().withRenderNameCase(RenderNameCase.LOWER)
+      .withInterpreterDialect(SQLDialect.POSTGRES)
+    val jdbcUrl = "jdbc:postgresql://localhost:5432/${baseTxnId}_${Random.nextLong()};DB_CLOSE_DELAY=-1;DATABASE_TO_LOWER=true"
+    val dataSource = PGSimpleDataSource()
+    dataSource.setURL(jdbcUrl)
+    val database = SqlProjectDatabaseImpl(dataSource, dialect = SQLDialect.POSTGRES)
+    projectFromXml(projectXml, baseTxnId) { database }
+    val clientConnection = dataSource.getConnection("postgres", "")
+    val serverConnection = dataSource.getConnection("postgres", "")
+    val serverDsl = DSL.using(serverConnection, settings)
+    val clientDsl = DSL.using(clientConnection, settings)
+    serverDsl.startTransaction()
+    clientDsl.startTransaction()
+    try {
+      serverTransaction.forEach {
+        it.colloboqueOperations.forEach {
+          LOG.debug("... applying operation={}", it)
+          serverDsl.execute(generateSqlStatement(serverDsl, it))
+        }
+      }
+      serverDsl.commit()
+      clientTransaction.forEach {
+        it.colloboqueOperations.forEach {
+          LOG.debug("... applying operation={}", it)
+          clientDsl.execute(generateSqlStatement(clientDsl, it))
+        }
+      }
+      clientDsl.commit()
+    } catch (e: Exception) {
+      LOG.info("Failed to execute transactions in parallel! Reason: ${e.localizedMessage}")
+      database.shutdown()
+      return false
+    }
+    database.shutdown()
+    return true
+  }
+
+  fun getProjectXml(projectRefid: ProjectRefid, baseTxnId: BaseTxnId): String {
```

Comment:
I suggest unifying this with `getActualSnapshot` method above. Add a second argument `baseTxnId?` to `getActualSnapshot` and use the maximum txn id if the passed one is null

Reply:
done